### ISSUE TITLE
CKAN: Datasets can be missed due to paging

### ISSUE THREAD

**Issue by kring** [TerriaJS/magda-metadata#26]
The CKAN Connector executes multiple `package_search` queries for a page of datasets at a time.  CKAN's [default search order](http://docs.ckan.org/en/latest/api/#ckan.logic.action.get.package_search) is first by relevance (presumably all datasets have equal relevance when there is no query) and second by modified date, descending.  This puts the most recently modified datasets at the beginning.

If new datasets are added while the crawler is running, they'll be missed, while other datasets will potentially be examined multiple times.  This is acceptable-ish because crawling is inherently a point-in-time activity.  The new datasets will be picked up on the next crawl.

More concerning, though, is if an existing dataset is _modified_ during crawling.  This will move it to the first page of results after the crawler has already moved to later pages.  Therefore, it will be missed.  Missing a dataset that existed prior to the start of crawling is not acceptable (-ish or otherwise).

It would be nice if CKAN datasets had a simple auto increment ID type of thing that we could use as the sort order, guaranteeing stable order and that new datasets added during crawling would be added at the end.  But since there is nothing like that (AFAICT), it's tricky.

One idea is to sort by ascending modified data, purposely request overlap in the pages, verify that overlap actually exists, and if it doesn't, re-request the page with an earlier start index until we find our overlap.

**Comment by tobybellwood** [comment]
Have we considered doing any smarts on the incremental package_search, using metadata_modified to order - and page until we meet a record already updated?  Whilst this may not solve up the issue, it will minimise the time it is missing, and minimise the chance of it occurring because of the reduced time querying?

This should enable more frequent increments (say hourly), coupled with a daily full refresh?

**Comment by kring** [comment]
yeah that's a good idea for doing incremental updates.  I believe CKAN also has an API for getting modified packages since a date.  But any kind of incremental update requires us to be able to do a full crawl and know that we got everything that was last modified before a certain date (the beginning or end of the crawl, let's say, depending on how the initial crawl works).  So this issue is about making sure the initial crawl is actually reliable.

**Comment by kring** [comment]
Although maybe what you're suggesting is that we make a full crawl reliable by doing it naively, and then following it up with incremental one(s) until we don't see any more changes.  That could work, and might be simpler than the page-with-overlap thing I described.

**Comment by tobybellwood** [comment]
yeah - if a full crawl is always followed by an incremental it should fix any updates.  The same logic may also be applied to "large" multi-page incrementals for surety?
